src/lib.rs

```
@@ -359,16 +398,33 @@ mod tests {
 
     #[test]
     fn test_parse() {
-        let input = b"\x00!B11:!B10;\x00\x00!\x00\x00\x00\x00";
+        let input = b"\x00!B11:!B10;\x00\x00!\x00\x00\x00\x00!B138";
         #[cfg(feature = "use_heapless")]
         let result = parse::<4>(input);
```

Was `MAX_RESULTS` set one higher than required intentionally? Eg. to make sure the parser doesn't accidentally return one more result than available? Then I'd say it should be updated to `5`.

(If I add another item to the input without updating the rest of the test case, `cargo test --features alloc` fails as expected, but `cargo test --features heapless` still succeeds, as the additional entry is silently ignored.)

---

there was no intention behind this with the test, i guess it should've been 3 so far.

you're right, in the `heapless` version we just return from `parse` once we hit `MAX_RESULTS` - in `alloc` we can continue until we're done.
i'm wondering if i'd need a way to signal this from within the parser to the consumer so that the consumer knows that there's more data to grab?

the more i'm thinking about this the more i think that the parser should maybe be implemented as an iterator. in that case i also wouldn't need the dependency to `heapless`. WDYT?